File: backend/src/main/java/com/dreampath/domain/chatbot/service/RagAnswerService.java
package com.dreampath.domain.chatbot.service;

import lombok.RequiredArgsConstructor;
import okhttp3.*;
import org.json.JSONArray;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class RagAnswerService {

    @Value("${openai.api.key}")
    private String openaiApiKey;

    private final OkHttpClient client = new OkHttpClient();

    /**
     * FAQ 전용 답변 생성 메서드
     */
    public String generateAnswer(String question, JSONArray matches) {
        // FAQ 형식으로 context 생성
        StringBuilder context = new StringBuilder();

        for (int i = 0; i < matches.length(); i++) {
            JSONObject m = matches.getJSONObject(i);

            if (m.has("metadata")) {
                JSONObject metadata = m.getJSONObject("metadata");

                if (metadata.has("question") && metadata.has("answer")) {
                    String faqQ = metadata.getString("question");
                    String faqA = metadata.getString("answer");
                    context.append("Q: ").append(faqQ).append("\n");
                    context.append("A: ").append(faqA).append("\n\n");
                } else if (metadata.has("text")) {
                    context.append("- ").append(metadata.getString("text")).append("\n");
                }
            }
        }

        String prompt = """
                당신은 DreamPath 진로 상담 서비스의 친절한 AI 어시스턴트입니다.
                아래 정보를 참고하여 사용자의 질문에 답변해주세요.
                친절하고 자연스럽게 한국어로 답변하세요.

                ⚠️ 중요 규칙:

                1. 먼저 참고 정보를 확인하세요:
                   - 참고 정보가 비어있거나 사용자 질문과 전혀 관련 없음 → "정보 없음 답변" 사용
                   - 참고 정보에 관련된 FAQ가 있음 → "FAQ 기반 답변" 제공

                2. 정보 없음 답변 (관련 FAQ가 없을 때):
                   반드시 정확히 이렇게만 답변:
                   "죄송하지만 해당 질문에 대한 정보가 없습니다. DreamPath팀에 문의를 남기시겠습니까?"

                3. FAQ 기반 답변 (관련 FAQ가 있을 때):
                   - FAQ 정보를 바탕으로 공감과 이해를 담아 대화하듯이 답변
                   - 사용자의 상황을 이해하고 있다는 느낌을 주세요
                   - 필요하다면 추가 팁이나 관련 정보를 덧붙이세요

                예시:
                - 좋은 예: "아이디 변경에 대해 문의주셨군요! 안타깝게도 DreamPath에서는 보안상의 이유로 한 번 설정된 아이디는 변경이 불가능해요."

                참고 정보:
                %s

                사용자 질문: %s
                """.formatted(context.toString(), question);

        try {
            JSONObject msg = new JSONObject()
                    .put("role", "user")
                    .put("content", prompt);

            JSONObject json = new JSONObject()
                    .put("model", "gpt-4o-mini")
                    .put("messages", new JSONArray().put(msg));

            RequestBody body = RequestBody.create(
                    json.toString(),
                    MediaType.parse("application/json")
            );

            Request request = new Request.Builder()
                    .url("https://api.openai.com/v1/chat/completions")
                    .addHeader("Authorization", "Bearer " + openaiApiKey)
                    .post(body)
                    .build();

            Response response = client.newCall(request).execute();
            String responseBody = response.body().string();

            JSONObject obj = new JSONObject(responseBody);
            return obj.getJSONArray("choices")
                    .getJSONObject(0)
                    .getJSONObject("message")
                    .getString("content");

        } catch (Exception e) {
            throw new RuntimeException("GPT 답변 생성 실패", e);
        }
    }
}

File: ai-service/services/chatbot/rag/rag_answer_service.py
import os
from openai import OpenAI
from typing import List, Dict, Any


class RagAnswerService:
    def __init__(self):
        self.client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

    def generate_answer(self, question: str, matches: List[Dict[str, Any]]) -> str:
        """FAQ 전용 답변 생성 메서드"""

        # FAQ 형식으로 context 생성
        context = []

        for match in matches:
            metadata = match.get("metadata", {})

            if "question" in metadata and "answer" in metadata:
                faq_q = metadata["question"]
                faq_a = metadata["answer"]
                context.append(f"Q: {faq_q}\nA: {faq_a}\n")
            elif "text" in metadata:
                context.append(f"- {metadata['text']}\n")

        # FAQ가 없으면 무조건 범위 외 메시지 반환
        if not context or len(context) == 0:
            return "죄송하지만, DreamPath 서비스 관련 질문 외에는 답변할 수 없습니다. DreamPath와 관련된 질문이나 도움이 필요하시면 언제든지 말씀해 주세요!"

        context_str = "\n".join(context)

        # FAQ가 있으면 GPT로 자연스러운 답변 생성
        prompt = f"""당신은 DreamPath의 비회원 전용 AI 챗봇입니다.

## DreamPath 서비스 범위 (비회원)
다음 주제에 **ONLY** 답변합니다:
- 서비스 소개 및 이용 방법
- 회원가입 안내
- 요금제 및 가격
- FAQ
- 문의하기 방법

## 절대 답변하지 않는 주제
- 일반 상식 (역사, 과학, 수학, 영어 등)
- 일상 대화 (날씨, 음식, 취미 등)
- 코딩/프로그래밍
- 의료/법률 상담
- 뉴스/시사
- 번역
- 기타 DreamPath 서비스 외 모든 질문

## 응답 규칙
1. **참고 정보와 관련된 질문 → 참고 정보 기반 답변**
   아래 참고 정보를 바탕으로 친절하게 답변하세요.

2. **서비스 외 질문 → 무조건 거절**
   반드시 다음 형식으로만 답변:
   "죄송하지만, DreamPath 서비스 관련 질문 외에는 답변할 수 없습니다. DreamPath와 관련된 질문이나 도움이 필요하시면 언제든지 말씀해 주세요!😊"

3. **응답 형식 → 마크다운**

참고 정보:
{context_str}

User Question: {question}

Answer (FAQ 정보를 기반으로 자연스럽게 답변):"""

        try:
            response = self.client.chat.completions.create(
                model="gpt-4o-mini",
                messages=[
                    {"role": "user", "content": prompt}
                ]
            )
            return response.choices[0].message.content

        except Exception as e:
            raise RuntimeError(f"GPT 답변 생성 실패: {str(e)}")
